### ISSUE TITLE
Offer simpler development workflow?

### ISSUE THREAD

**Issue by astrojuanlu** [scipy/scipy#12633]
I find the reliance on conda in the Ubuntu development environment quickstart guide exceedingly complex, not to mention the need to rename `libgfortran.so`, but I guess that those instructions are there for a good reason (historical pain setting up development environments, messed up Python installations... and other incarnations of Evil :tm: ). However, I tried this on a modern Linux Mint 20 box (based on Ubuntu 20.04 LTS Focal Fossa) and this seemed to work just fine:

```
$ # create and activate pyenv/venv/virtualenv environment
(venv) $ pip install -e .
(venv) $ pip install numpy  # Surprisingly, NumPy didn't get installed in this step
(venv) $ pip install pytest
(venv) $ pytest -k optimize  # Works!
```

Is there any potential drawback in following this method?

**Comment by mdhaber** [comment]
I don't use pip, so I can't comment on that procedure, but you're right that you don't have to use conda. I use conda, and the guide reflects what I needed to do to get conda to work on WSL reliably. Yes, the renaming of `libgfortran.so` has been a sore spot, and some people are fortunate not to need it, but it is a fix to a common issue. 

I'll try to post a PR for a pip guide today.

**Comment by astrojuanlu** [comment]
Thanks for your quick response @mdhaber!

> when do pybind11 and Cython get installed in those four lines?

It's exactly in `pip install -e .`.

> It looks like you already have all the compilers?

Yes.

---

This is a way to reproduce the "simplified" installation procedure:

```
$ docker run -it --rm ubuntu:20.04 bash
# apt update && apt install -y python3-venv python3-dev git libopenblas-dev gfortran g++
# useradd dummyuser && su - dummyuser  # Avoid running as super user
$ python3 -m venv /tmp/venv && . /tmp/venv/bin/activate
(venv) $ python -m pip install -U pip setuptools wheel  # Upgrade to latest versions
(venv) $ cd /tmp && git clone https://github.com/scipy/scipy.git
(venv) $ cd scipy && python -m pip install -e .  # Works!
(venv) $ python -m pip list  # Why no numpy?
Package       Version            Location
------------- ------------------ ----------
pip           20.2
pkg-resources 0.0.0
scipy         1.6.0.dev0+8692c88 /tmp/scipy
setuptools    49.2.0
wheel         0.34.2
(venv) $ python -m pip install numpy  # Shouldn't be needed...
(venv) $ python -m pip install pytest && pytest  # Also works!
```

**Comment by ev-br** [comment]
FWIW, if the goal has the word "simple" in it, it shouldn't involve docker :-). It should however mention that compilers need to be sudo apt-get install-ed, while sudo pip in a big no-no. 

As long as it's an alternative to more detailed instructions for conda, pip instructions can be higher level.

**Comment by mdhaber** [comment]
@astrojuanlu I am going to to do a fresh install of Windows Subsystem for Linux with Ubuntu 16. What commands would you suggest to go from scratch to a development enviroment?

**Comment by astrojuanlu** [comment]
> FWIW, if the goal has the word "simple" in it, it shouldn't involve docker :-)

Of course, Docker was only to demonstrate the commands in a freshly installed machine :) It shows the bare minimum to get up and running.

> I am going to to do a fresh install of Windows Subsystem for Linux with Ubuntu 16. What commands would you suggest to go from scratch to a development enviroment?

Ubuntu 16 shipped Python 3.5 by default, so some extra work is needed to install a more recent Python version, for example 3.6:

```
sudo apt update && sudo apt install software-properties-common
sudo add-apt-repository ppa:deadsnakes/ppa
sudo apt update && sudo apt install python3.6 python3.6-dev python3.6-venv
```

From here, these commands should work (assuming WSL runs as a non-privileged user, the shell is Bash, and `sudo` exists):

```
sudo apt update
sudo apt install -y git libopenblas-dev gfortran g++
python3.6 -m venv /tmp/venv
source /tmp/venv/bin/activate
python -m pip install -U pip setuptools wheel
git clone https://github.com/scipy/scipy.git
cd scipy
python -m pip install -e .
```

These are more or less a translation of what I already tried in Ubuntu 20.04, on Docker everything seems to work.

**Comment by peterbell10** [comment]
>Curious though - when do pybind11 and Cython get installed in those four lines?

I think pip reads the `pyproject.toml` file and installs all dependencies listed there.

**Comment by astrojuanlu** [comment]
Yes, `[build-system]` in `pyproject.toml` (aka [PEP 518](https://www.python.org/dev/peps/pep-0518/)) is the replacement of the old, never-quite-working `setup_requires` of setuptools. It's interesting that `numpy` doesn't seem to get listed in `install_requires` though (PEP 518 and previously `setup_requires` are about build requirements, `install_requires` is about runtime requirements).

**Comment by astrojuanlu** [comment]
Also, notice that in the PEP 518 world, the dependencies of the `[build-system]` are installed in an isolated environment that is only created temporarily to build SciPy, but are not installed in the actual environment. (Sorry for the small digression, hope it clarifies things)

**Comment by mdhaber** [comment]
Turns out WSL has upgraded to 20.04, so I did:
```
sudo apt update
sudo apt install -y git libopenblas-dev gfortran g++
python3.6 -m venv /tmp/venv
```
After an error about Python (because it is in fact 20.04 instead of 16.04)
```
sudo apt install -y python3-venv
python3.6 -m venv /tmp/venv
source /tmp/venv/bin/activate
python -m pip install -U pip setuptools wheel
git clone https://github.com/mdhaber/scipy.git
cd scipy
git remote add upstream https://github.com/scipy/scipy.git
git fetch upstream
git rebase upstream/master
python -m pip install -e .
```
I assume this can be consolidated by just adding python3-venv in the first `sudo apt install` line.

But build fails ending in:
```
    error: Command "x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I/home/mdhaber/scipy/scipy/special -I/tmp/pip-build-env-fw66g5co/overlay/lib/python3.8/site-packages/numpy/core/include -I/usr/include/python3.8 -Iscipy/_lib -Iscipy/_build_utils/src -I/tmp/pip-build-env-fw66g5co/overlay/lib/python3.8/site-packages/numpy/core/include -I/tmp/venv/include -I/usr/include/python3.8 -I/tmp/venv/include -I/usr/include/python3.8 -c scipy/special/cephes/dawsn.c -o build/temp.linux-x86_64-3.8/scipy/special/cephes/dawsn.o -MMD -MF build/temp.linux-x86_64-3.8/scipy/special/cephes/dawsn.o.d" failed with exit status 1
    ----------------------------------------
ERROR: Command errored out with exit status 1: /tmp/venv/bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/home/mdhaber/scipy/setup.py'"'"'; __file__='"'"'/home/mdhaber/scipy/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' develop --no-deps Check the logs for full command output.
```

**Comment by mdhaber** [comment]
Interestingly, I was able to build with conda. After all that, I just did:
```
rm -r build
cd ..
curl -O https://repo.anaconda.com/archive/Anaconda3-2020.07-Linux-x86_64.sh
bash Anaconda3-2020.07-Linux-x86_64.sh
(source .bashrc or restart console)
conda create --name scipydev python=3.8 cython numpy pytest pybind11
conda activate scipydev
cd scipy 
python setup.py build_ext --inplace
conda develop .
```
So maybe 20.04 users can skip the Homebrew on Linux compilers and libgfortran.so renaming. Would that help?

**Comment by andyfaff** [comment]
The CI scripts ("nightly GH actions build) and the Dockerfile probably contain the best setup template for a clean system. I.e. what do I have to run to get stuff to build. I'm talking about the lines in them, not running docker itself.

**Comment by astrojuanlu** [comment]
@mdhaber Weird, I can't imagine how the compilation of `dawsn.c` could have failed.

I confirm that [your instructions](https://github.com/scipy/scipy/issues/12633#issuecomment-665914296) work in a fresh Ubuntu 20.04 Docker image:

```
$ docker run -it --rm ubuntu:20.04 bash
# apt update && apt install -y curl git libopenblas-dev gfortran g++
# curl -O https://repo.anaconda.com/archive/Anaconda3-2020.07-Linux-x86_64.sh
# bash Anaconda3-2020.07-Linux-x86_64.sh
# exec $SHELL
# conda create --name scipydev python=3.8 cython numpy pytest pybind11
# conda activate scipydev
(scipydev) # git clone https://github.com/scipy/scipy.git
(scipydev) # cd scipy
(scipydev) # python setup.py build_ext --inplace
```

But again, I fail to see what's the contribution of Anaconda here. The compiler was installed with `apt` after all. The implicit `conda install cython numpy pytest pybind11` is not necessary for me, `pip install -e .` works inside the conda environment.

**Comment by mdhaber** [comment]
> But again, I fail to see what's the contribution of Anaconda here.

I wouldn't have expected one either. But one built and one didn't. I can try it again later, but I'm not going to get to simplified instructions today after all. 

> `pip install -e .` works inside the conda environment

If possible, I try to stick to conda in a conda environment. I've been told they don't play as well together as they should.

**Comment by astrojuanlu** [comment]
You're right. I tested `conda develop .` though, and it works! 🚀

**Comment by ev-br** [comment]
BTW, where is the quickstart tuturial? I failed to find it on scipy.org, and googling around brings numpy quickstart tutorial instead.

**Comment by ilayn** [comment]
It is indeed very difficult to navigate in the docs for these pages 

http://scipy.github.io/devdocs/dev/contributor/contributor_toc.html

**Comment by ev-br** [comment]
Thanks Ilhan! I wonder what does it take to make this part of the docs more discoverable. 

Here's a quick stab: gh-12649 . Like the PR says, I do not insist on the choices made for various steps, this is basically a recording of what I just did after reinstalling the system on a dev laptop. The level of detail is also subjective: this is what I as user would be most happy with, but I may or may not be a good proxy for a target audience. 
Feel free to run with it, rewrite it or do it some other way :-).

**Comment by mdhaber** [comment]
> It is indeed very difficult to navigate in the docs for these pages

It is directly linked from the main documentation page.
http://scipy.github.io/devdocs/
Under Developer Documentation. It's the third link.

**Comment by ev-br** [comment]
![Screenshot_20200731-180040~2](https://user-images.githubusercontent.com/2133832/89048422-79119680-d358-11ea-81e0-93c3a31ed585.png)

Maybe it can be made more googlable? It did not immediately show in the "scipy dev guide" or "scipy startup guide" or "scipy quickstart" searches for me.

**Comment by ilayn** [comment]
There are too many developer and contribute words in this page. It is creating a mental load when you skim over the links to slow down and end up on your target. Repetition of module names in bold above (tutorials) and below (API) is also breaking a few well-established UX rules too. Then there is the right hand side which links to the things you see on the left. Also there is an extra API link at the top. It is all very confusing for passing-by people and not the easiest for experienced people. 

So I think we can improve these things a lot. However it is not a trivial job either. NumPy folk got huge success out of the doc project so maybe we can reap the low-hanging fruits based on their experience.

**Comment by rgommers** [comment]
For the `docs.scipy.org` front page, we should completely redo it, or better - we won't need it anymore once we get the new Sphinx theme going. That'll have a version switcher dropdown, and NumPy releases can disappear now that NumPy hosts their own. So then docs.scipy.org can drop straight into the docs for the latest stable release.

**Comment by WillTirone** [comment]
Just wanted to say I definitely agree on the organization of things - I had quite the time getting things up and running, but I can definitely see it being discouraging for people starting. I think this is a helpful change!

**Comment by rgommers** [comment]
I think yes, we need to revisit this and address many of the points brought up here. And shorten every page, there's a lot of repetition and content that isn't needed and just scares away people. Things should work out of the box fairly easily; the new `pip` page shows that. For `conda` it's easier than for pip, especially on Linux. All that is needed is:
```
# Execute this from the root of the scipy repo, with a vanilla Miniconda/Anaconda install
conda create -n scipy-dev python=3.8 numpy pybind11 cython pytest gfortran_linux-64 gxx_linux-64 mypy sphinx
conda activate scipy-dev
python runtests.py
```
All the `numpy.distutils` and conda compilers issues have been addressed AFAIK.

**Comment by mdhaber** [comment]
I am the author of the existing quickstart guides. I know they're fun to criticize, but when you do set about to improve them, please keep the following in mind.
a) I tried dozens of build variations on several computers and settled on what is in the guides because they worked on all of those computers. Please simplify if possible, but please test your procedure on many machines.
b) When I started contributing to SciPy, I didn't know a fork from a branch. As you strip these down, please consider whether it is more difficult for the beginner to find the information that isn't there or for the expert to skip over what is.

After following steps 1-5 and 8-14 of the Ubuntu guide on a fresh EC2 Ubuntu 18.04 instance - because remember, you have to install Anaconda, create an environment, install Git, and clone the SciPy repo - and the three lines above worked. @rgommers It does seem that the conda compiler issues have finally been fixed, so maybe steps 7, 16, and 17 can be eliminated. I hated having to include them in the first place, too.

**Comment by rgommers** [comment]
> I worked with an absolute beginner to test and refine the Ubuntu guide before it was merged. Consider doing the same, as decades of experience tends to change one's perception of what belongs and what doesn't.

That's not quite it. The problem is that multiple things are meshed together, and then repeated multiple times in various guides. It's a massive wall of text, which is discouraging. A simple _"Prerequisites are git and a working conda install. If you don't have git, see HERE, ..."_. And same with explain how the very basic workflow in GitHub works.

**Comment by WillTirone** [comment]
> I worked with an absolute beginner 
I'm mostly an absolute beginner, and I've been looking for easy things to work on with SciPy - let me know if I can assist. I have a working build although haven't made any commits to the code base but I'm happy to help with this issue if I can

**Comment by oulenz** [comment]
> I think yes, we need to revisit this and address many of the points brought up here. And shorten every page, there's a lot of repetition and content that isn't needed and just scares away people. Things should work out of the box fairly easily; the new `pip` page shows that. For `conda` it's easier than for `pip`, especially on Linux. All that is needed is:
> 
> ```
> # Execute this from the root of the scipy repo, with a vanilla Miniconda/Anaconda install
> conda create -n scipy-dev python=3.8 numpy pybind11 cython pytest gfortran_linux-64 gxx_linux-64 mypy sphinx
> conda activate scipy-dev
> python runtests.py
> ```
> 
> All the `numpy.distutils` and conda compilers issues have been addressed AFAIK.

I'm about to write my first PR for scipy, and I'm really glad I found this. I was particularly confused by the fact that there are two setup instructions for Ubuntu listed [here](http://scipy.github.io/devdocs/dev/contributor/contributor_toc.html#contributor-toc). The first one didn't look right because it doesn't mention conda, the second one didn't look right because it seems to explicitly target Ubuntu 16.04, and because it somehow involves homebrew.

So yes, please update and simplify the instructions.

**Comment by mdhaber** [comment]
@oulenz the update PR gh-12695 is waiting to be merged.

> seems to explicitly target Ubuntu 16.04, and because it somehow involves homebrew

For a time, homebrew was the simplest option that was robust, and at that time, the instructions had only been validated for 16.04.

**Comment by rgommers** [comment]
I came across this problem again today. Had to find instructions for a colleague to get started. They know what they're doing, have conda installed, etc. - we have a ton of docs to wade through, and exactly zero contain just the commands to get started without two pages of text. This is what I wanted, in the most easy-to-find place:

```
# Best practice, use an environment rather than install in the base env
conda create -n my-env
conda activate my-env

# Prefer conda-forge (because it has the compilers we need)
conda config --env --add channels conda-forge

# Now install all packages that are needed in one `conda install` command
conda install python=3.8 numpy pybind11 cython pytest sphinx matplotlib mypy compilers

# Build SciPy for development work plus run tests
python runtests.py    # Alternatively, it's fine to use `python setup.py develop`
```

I'll mark this for the next release. I'll try and submit a PR for this next month.

**Comment by mdhaber** [comment]
Or I can post it today.

**Comment by tupui** [comment]
The organisation will slightly change with the new theme #13724.

**Comment by tupui** [comment]
Not sure if relevant here. But, we could add a section about CI and how to run all tools locally. This is something I am often in need and I resort most of the time to reading the different yaml files. What do you think? We could also talk about how to skip CI.

**Comment by rgommers** [comment]
A "Continuous Integration" section in http://scipy.github.io/devdocs/dev/contributor/contributor_toc.html seems justified. Explaining the basics of what is run where, under what circumstances `[ci skip]` is okay, and what to do if something fails.

**Comment by ilayn** [comment]
CI is for us maintainers mostly. People often make a one-off PR as they see fit and won't be bothered with CI anyways. So probably we overwhelm them with too many details.

**Comment by tupui** [comment]
Then in the maintainer's doc? This still has some value for newcomers like me for instance ;)

**Comment by ilayn** [comment]
You have all the rights to be unsure :) because it doesn't work properly across the board and Azure is notoriously resisting to implement one. There is an Azure Pipelines issue with almost a thousand people interacted with them but MS is being MS.

**Comment by mdhaber** [comment]
> So probably we overwhelm them with too many details.

That's something to consider, but I think @rgommers makes a good point that a CI section would be nice to have - even for new-contributors. Seeing lots of errors on a PR that they didn't see on their own computers can be overwhelming, too. It would be helpful write down some strategies for how to find and read the messages. This would eventually save maintainers some time, so that they don't need to point out similar things PR after PR (e.g. find the test failure messages at the end, look for warnings in the doc build log, etc.).

So I agree that it's important to present them in a way that doesn't overload beginners, but I do think such a guide would be helpful even for beginners.

**Comment by tupui** [comment]
Looking at the ultra-quick start guide (conda on Mac/unix), we might be missing `conda develop .` no?

**Comment by rgommers** [comment]
Indeed

**Comment by tupui** [comment]
Ok then I will write some bytes about CI and this at the same time.

**Comment by tylerjereddy** [comment]
I think some of this is actually in `1.7.0` but the issue is still open, I'm not too worried about it though

**Comment by melissawm** [comment]
Ok folks - I am hijacking this issue for a bit but I think this discussion fits here. Currently, this is where you can find building instructions for SciPy:

- Building index: https://scipy.github.io/devdocs/building/index.html
- Building FAQ: https://scipy.github.io/devdocs/building/faq.html

(ordered by workflow)
- pip+venv+meson+linux → [quickstart_pip](https://scipy.github.io/devdocs/dev/contributor/quickstart_pip.html)
- pip+virtualenvwrapper+meson+linux → [quickstart_pip](https://scipy.github.io/devdocs/dev/contributor/quickstart_pip.html)
- conda+setuptools+linux → [quickerstart_conda](https://scipy.github.io/devdocs/dev/contributor/quickerstart_conda.html), [quickstart_ubuntu](https://scipy.github.io/devdocs/dev/contributor/quickstart_ubuntu.html)
- conda+setuptools+mac → [quickerstart_conda](https://scipy.github.io/devdocs/dev/contributor/quickerstart_conda.html), [quickstart_mac](https://scipy.github.io/devdocs/dev/contributor/quickstart_mac.html)
- pip+setuptools+windows → [building/windows](https://scipy.github.io/devdocs/building/windows.html)
- docker+conda+setuptools → [quickstart_docker](https://scipy.github.io/devdocs/dev/contributor/quickstart_docker.html)
- gitpod+conda+setuptools → [quickstart_gitpod](https://scipy.github.io/devdocs/dev/contributor/quickstart_gitpod.html)
- setuptools+linux → [building/linux](https://scipy.github.io/devdocs/building/linux.html)
- setuptools+mac → [building/macosx](https://scipy.github.io/devdocs/building/macosx.html) (outdated, with a warning)

I do have ideas to simplify these pages, and help guide users through them, but I wonder 

1. what workflows do we want to keep supporting even after the meson switch is completed
2. how this relates to the work being done towards #15489

**Comment by ilayn** [comment]
Oh my. That's an amazing and terrifying list simultaneously. Thanks so much for this @melissawm! 

I would throw the following tree idea just to keep the discussion rolling otherwise things get very abstract very quickly. 

```
Instruction landing page

|__ pip based 
|     |
|     |__ Linux with/out venv
|     |__ Win with/out venv
|     |__ Apple with/out venv
|__ conda based 
|     |
|     |__ Linux
|     |__ Win
|     |__ Apple
|__ docker
|__ gitpod
|__ Miscellaneous
```

We probably won't need `setuptools`  ones very soon as we are trying to unify things.

**Comment by ev-br** [comment]
A couple of idle thoughts:

- please don't advertise anything without a virtual environment (pip or conda, no matter) 
- maybe simplify the number of pages down to : a very quick start (both pip and conda, two sets of instructions in one page), ending with a dev.py call + a manual build showing meson incantations + a sepate widows page 

Note that Mac OS and Linux are actually very similar and for a user who is not very novice, the difference is likely rather minor (eg, the second page can use either Mac OS Linux).

**Comment by stefanv** [comment]
Agreed with @ev-br's comments. For developers, it may be sufficient to start with the requirement that you are working in a virtualenv (we can teach that elsewhere and point to it).  Then, at least, we only have to deal with two "system" scenarios, one in which you need to install non-Python dependencies manually and the other where they come in via conda-forge.

Emphasizing another point from above: it is important to have the ability to install a built development version so that it can easily be experimented with in IPython or similar.

**Comment by rgommers** [comment]
Thanks @melissawm & all! I agree with most/all of what's being suggested above. One important point I want to add: we should separate environment setup from the actual SciPy build & develop steps. The latter (`git` / `runtests.py` / `dev.py` / `pytest` / ...) are 100% identical across pip/conda and Windows/Linux/macOS. So why do we keep repeating those over and over?

Instead, let's teach the user how to think about the various parts here, and then have platform and package manager specific instructions only where they are actually different.

I think we have two special cases in the current docs: Gitpod and Docker. For the rest, it all fits a pattern like:

1. Obtain source code (clone git repo if you don't have it yet)
2. Set up a development environment (conda, or some flavor of the many virtualenv management tools)
  2b. Explain that for virtualenv we cannot get all dependencies, so use apt/yum/homebrew/etc. to obtain the missing ones
3. Build SciPy
4. Perform development tasks: run tests, benchmarks, etc. 

And then we also pick one way of doing things per platform + package manager. For example, the Windows + venv docs now have a huge section on building OpenBLAS from source. This should be moved out, there's no point asking a SciPy contributor to do that. Instead, we just point them to our preferred binary and are done with it.

It's worth some thought about whether we can use a widget-type thing to select instructions per platform and pip/conda, then it perhaps could all fit on a single html page?

**Comment by ilayn** [comment]
One detail I am not really sure yet is that if windows user doesn't build openblas the downloaded one might have ABI incompatibilities during the compilation of fortran parts. That **was** the reason of that section. I'm not sure if that still is the case though. If not I agree that is just plain agony to build openblas.

**Comment by rgommers** [comment]
> One detail I am not really sure yet is that if windows user doesn't build openblas the downloaded one might have ABI incompatibilities during the compilation of fortran parts.

That's only true if that user gets their Fortran compiler from some random place. We should be recommending compilers and an OpenBLAS binary that are compatible. If we can do it in CI, then a user can do it as well.

**Comment by melissawm** [comment]
> Thanks @melissawm & all! I agree with most/all of what's being suggested above. One important point I want to add: we should separate environment setup from the actual SciPy build & develop steps. The latter (`git` / `runtests.py` / `dev.py` / `pytest` / ...) are 100% identical across pip/conda and Windows/Linux/macOS. So why do we keep repeating those over and over?
> Instead, let's teach the user how to think about the various parts here, and then have platform and package manager specific instructions only where they are actually different.

Absolutely agree :+1:

> It's worth some thought about whether we can use a widget-type thing to select instructions per platform and pip/conda, then it perhaps could all fit on a single html page?

I have just the thing in mind: https://github.com/executablebooks/sphinx-tabs\

You can see it in action here, for example:

https://napari.org/plugins/guides.html#reader-example

**Comment by melissawm** [comment]
> One detail I am not really sure yet is that if windows user doesn't build openblas the downloaded one might have ABI incompatibilities during the compilation of fortran parts. That **was** the reason of that section. I'm not sure if that still is the case though. If not I agree that is just plain agony to build openblas.

Could we make the dev guides lighter and add those gotchas to the FAQ instead? It is almost empty but seems to be the best place for special cases.

**Comment by ilayn** [comment]
I think we could. Back then we had so many issues with different people working with different compilers and also hard time picking the right one depending on how they installed them etc. But I think things are more stabilized now for discovering the right ones so we can be less paranoid in the language.

**Comment by melissawm** [comment]
I have a draft that I am almost ready to submit, but would like some feedback on two specific pages. Just checking if we need to keep this content, update it or maybe just remove it entirely. The two I'm thinking of are

- MacOS https://scipy.github.io/devdocs/building/macosx.html
- These instructions for MKL https://scipy.github.io/devdocs/building/linux.html#intel-c-compiler-and-mkl

The way I'm setting this up is that all the system-level dependencies are one step (with descriptions for all three Mac, Windows and Linux) and the python setup which is pip or conda specific is another step. This should drastically reduce duplication. We can still have complete walkthroughs for each of the most common workflows but then we'd have a much lower maintenance surface.

**Comment by rgommers** [comment]
@melissawm those macOS and MKL instructions are both very outdated and are fine to remove.

> On a mac, I've installed iterm2, git through homebrew, and (mini)conda

I would avoid iterm2, because the built-in Terminal app is perfectly fine. Git through Homebrew is the best idea indeed. And I'd say Miniforge is recommended over Miniconda. Actually, while we're at it, let's recommend Mambaforge (https://github.com/conda-forge/miniforge#mambaforge). That has the extra bits people are going to want when the going gets slow.

**Comment by tupui** [comment]
> @melissawm those macOS and MKL instructions are both very outdated and are fine to remove.
> 
> > On a mac, I've installed iterm2, git through homebrew, and (mini)conda
> 
> I would avoid iterm2, because the built-in Terminal app is perfectly fine. Git through Homebrew is the best idea indeed. And I'd say Miniforge is recommended over Miniconda. Actually, while we're at it, let's recommend Mambaforge ([conda-forge/miniforge#mambaforge](https://github.com/conda-forge/miniforge#mambaforge)). That has the extra bits people are going to want when the going gets slow.

+100 for `mambaforge`. It might just be hard for newcomers to quick start though due to the need to use both `conda` and `mamba` commands. With the new solver coming up to `conda`, do we actually need to recommend `mambaforge`?

**Comment by andyfaff** [comment]
Why not install git from conda-forge? Then there's no need to install homebrew.

**Comment by astrojuanlu** [comment]
The new conda libmamba solver is indeed very fast, but it's behind and experimental flag.

**Comment by rgommers** [comment]
> Why not install git from conda-forge? Then there's no need to install homebrew.

The problem with that is that it goes into some environment. Installing anything into the `base` environment is not recommended (it's too easy to break `base` that way), and installing into any other environment means it's not always available on the command line. Best practice is to install standalone command-line tools like `git`, `ripgrep`, `gh`/`hub`, `rsync`, etc. with Homebrew, and anything that's part of the Python development process (compilers, openblas, etc.) in conda environments.

**Comment by andyfaff** [comment]
Definitely agree it shouldn't be installed into `base`. One could argue that installing those standalone CLT with homebrew is like meddling with a base-type environment. If compilers/openblas are going to be installed into a dev-type conda environment, why should it be any different installing those CLT tools into the same env? It could make things a lot cleaner.

It would also simplify instructions, i.e:

- install conda (/mamba/whatever)
- use conda commands to setup a dev env with CLT/compilers

Instead of:

- install conda
- install brew
- use brew to install CLT
- use conda command to setup dev env with compilers

**Comment by eli-schwartz** [comment]
> For example, the Windows + venv docs now have a huge section on building OpenBLAS from source. This should be moved out, there's no point asking a SciPy contributor to do that. Instead, we just point them to our preferred binary and are done with it.

> One detail I am not really sure yet is that if windows user doesn't build openblas the downloaded one might have ABI incompatibilities during the compilation of fortran parts. That **was** the reason of that section. I'm not sure if that still is the case though. If not I agree that is just plain agony to build openblas.

If someone has a bunch of spare time on their hands, it may be worth considering to port openblas to Meson also, and then using that as a Meson subproject dependency fallback. You wouldn't even need to document it. :)

That would make it essentially transparent to users as an alternative to suggesting that they build it anyway, but do so manually.

**Comment by rgommers** [comment]
> One could argue that installing those standalone CLT with homebrew is like meddling with a base-type environment. If compilers/openblas are going to be installed into a dev-type conda environment, why should it be any different installing those CLT tools into the same env? It could make things a lot clean

Homebrew has a very different design, there's no base + activated envs. `git` is also something almost everyone (except absolute beginners) will already have, so it'd be bad to add it to every conda env - it makes conda solves slower, and then you just have multiple copies of `git` lying around. I'd also be fine with not having any instructions for it beyond mentioning it's needed.

> If someone has a bunch of spare time on their hands, it may be worth considering to port openblas to Meson also, and then using that as a Meson subproject dependency fallback. You wouldn't even need to document it. :)

I like the `subproject` idea, but this port would not be a small undertaking. And it's fragile unless upstream is happy with either switching to Meson or maintaining two build systems in parallel. Alternatively, is it possible to use a subproject and let Meson drive the build through CMake? https://github.com/xianyi/OpenBLAS is full of Makefile's but CMake seems to be supported directly as well.

**Comment by eli-schwartz** [comment]
Meson has tier two support for cmake subprojects. It assumes that cmake is installed as an additional build dependency for Meson projects, which is not always true, and relies on using cmake's trace API which is then parsed and dynamically converted into native Meson rules.

It may or may not work. The hope is that it does, and many projects do build successfully that way... but also, many do not. The more complicated the cmake usage is, the more likely it will hit an edge case that Meson doesn't handle.

Try it and see if it works. :) If it doesn't work, we will try to fix it.

(Just to be clear, by "we" I mean the one person on the Meson dev team that added the cmake support and is in charge of making it work. A very dedicated person, to be sure -- who else would dive that deeply into cmake? :D)

**Comment by eli-schwartz** [comment]
As far as upstream support goes, they support two build systems right now already.

There's a related lapack issue where lapack had a meson.build, but removed it because it was PRed by one person in an incomplete stage, apparently doesn't work well, and the original submitter disappeared immediately. It wasn't documented in the README and most people didn't know it existed.

They still kept both the Makefiles and cmake because, apparently, people are actively using it and it's popular. This despite a bunch of people who dislike anything other than "good old makefiles" who have "learned to tolerate cmake because it helps other people" or some such.

So... It's possible that if SciPy says "we would find meson support really helpful and are willing to help maintain it in the long term" then they would accept that. On the other hand, maybe not. I don't know.

The word of a significant and active project planning to use an openblas meson port as core infrastructure, *does* carry more weight than a lone-dev one-off code dump at least. So there is that.

**Comment by rgommers** [comment]
Yes, we have a good relationship with OpenBLAS so if we'd contribute the build support and promise some help going forward, it'll probably be accepted. Let's see if we have a volunteer for this project.

**Comment by ilayn** [comment]
As far as I can tell OpenBLAS has most of the machinery in place (probably painstakingly) developed over years. As far as I can tell, it works quite well even on Windows/MinGW with relatively low maintenance (except where the machine architecture selection goes awry). But typically quite painless. So it might even be a simple cmake trigger through meson. 

But we are all Python-based users so I'm not familiar with other lang use-cases and whether this might be an option. @martin-frbg typically has wise things to say on the subject.

**Comment by martin-frbg** [comment]
Do you happen to have some pointers for meson driving a build through cmake ? The cmake support in OpenBLAS is a relatively recent addition (and still carries a warning that it may not be exactly equivalent to a gmake build). Some of its functions are a bit involved as they have to do weird things like parsing conditional code in files with Makefile syntax. On the other hand I am not sure I (or anybody) would be happy with supporting three build systems in parallel. (Even Reference-LAPACK has unresolved discrepancies between their gmake and cmake setups, which was part of their reason for throwing out meson one version after accepting that code drop)

**Comment by eli-schwartz** [comment]
Here are the docs on Meson's support for cmake subprojects: https://mesonbuild.com/CMake-module.html#cmake-subprojects

Using it is simple enough, as long as it works correctly.

**Comment by martin-frbg** [comment]
btw regarding compilation of the "fortran parts" and potential ABI incompatibilities - I have a half-baked PR going that (optionally) uses f2c-mangled versions of said parts. I "only" need to track down test errors in all data type flavors of ?GELSD
